src/xpcc/driver/position/vl53l0_impl.hpp

```diff
@@ -0,0 +1,641 @@
+	VL53L0_RF_CALL(write(Register::SYSTEM__INTERRUPT_CONFIG_GPIO, InterruptConfig::NewSampleReady));
+	VL53L0_RF_CALL(updateControlRegister(Register::GPIO__HV_MUX_ACTIVE_HIGH,
+										 Control_t(0), GpioConfig::InterruptPolarityHigh));
+	VL53L0_RF_CALL(write(Register::SYSTEM__INTERRUPT_CLEAR, InterruptClear::Range));
+
+	// set default measurement sequence (TCC and MSRC disabled)
+	VL53L0_RF_CALL(write(Register::SYSTEM__SEQUENCE_CONFIG, standardSequence));
+
+	// recalculate measurement timings
+	VL53L0_RF_CALL(setMaxMeasurementTime(this->measurementTimeUs));
+
+	// VHV calibration
+	VL53L0_RF_CALL(write(Register::SYSTEM__SEQUENCE_CONFIG, MeasurementSequenceStep::VhvCalibration));
+	if(not RF_CALL(performReferenceCalibration(Start::VhvCalibrationMode))) {
+		XPCC_LOG_ERROR << "VHV calibration failed." << xpcc::endl;
+		RF_RETURN(false);
+	}
+
+	// phase calibration
+	VL53L0_RF_CALL(write(Register::SYSTEM__SEQUENCE_CONFIG, MeasurementSequenceStep::PhaseCalibration));
+	if(not RF_CALL(performReferenceCalibration())) {
+		XPCC_LOG_ERROR << "Phase calibration failed." << xpcc::endl;
+		RF_RETURN(false);
+	}
+
+	// restore measurement sequence settings
+	VL53L0_RF_CALL(write(Register::SYSTEM__SEQUENCE_CONFIG, standardSequence));
+
+	RF_END_RETURN(true);
+}
+
+template < typename I2cMaster >
+xpcc::ResumableResult<bool>
+xpcc::Vl53l0<I2cMaster>::loadTuningSettings()
+{
+	using namespace vl53l0_private;
+
+	RF_BEGIN();
+
+	// write the configuration
+	for(index = 0; index < 80; index++)
+	{
+		i2cBuffer[1] = configuration[index].value;
+		VL53L0_RF_CALL( write(Register(configuration[index].reg), i2cBuffer[1]) );
+	}
+
+	RF_END_RETURN(true);
+}
+
+template < typename I2cMaster >
+xpcc::ResumableResult<bool>
+xpcc::Vl53l0<I2cMaster>::readSensor()
+{
+	static_assert(MaxMeasurementTimeUs / 1000 * 2 <= ArithmeticTraits<uint16_t>::max,
+				"MaxMeasurementTimeUs out of range");
+
+	RF_BEGIN();
+
+	data.reset();
+
+	// undocumented black magic
+	VL53L0_RF_CALL(write(Register(0x80), 0x01));
+	VL53L0_RF_CALL(write(Register(0xFF), 0x01));
+	VL53L0_RF_CALL(write(Register(0x00), 0x00));
+	VL53L0_RF_CALL(write(Register(0x91), stopMode));
+	VL53L0_RF_CALL(write(Register(0x00), 0x01));
+	VL53L0_RF_CALL(write(Register(0xFF), 0x00));
+	VL53L0_RF_CALL(write(Register(0x80), 0x00));
+
+	// start range measurement
+	VL53L0_RF_CALL(write(Register::SYSRANGE__START, Start::StartStop));
+
+	// wait for the start flag to be cleared, use default timeout
+	VL53L0_RF_CALL(poll(Register::SYSRANGE__START, [](uint8_t value) {
+		return not (value & uint8_t(Start::StartStop));
+	}));
+
+	// wait for the measurement to finish, timeout of 2 * this->measurementTimeUs
+	VL53L0_RF_CALL(poll(Register::RESULT__INTERRUPT_STATUS, [](uint8_t value) {
+		return value & (InterruptStatus::NewSampleReady | InterruptStatus::OutOfWindow).value;
+	}, measurementTimeUs / 1000 * 2));
+
+	// read 16 bit range value
+	VL53L0_RF_CALL(read(Register::RESULT__RANGE_VALUE0, &data.distanceBuffer[0], 2));
```

Comment:
Careful, this can lead to data "tearing" issues, since the I2C operates asynchronously. If you call `getData().isValid()` from a protothread while running this function from another, you may catch the case where just the first byte is updated which may give you false positives.

Using the `i2cBuffer` and then copying it over "proto-atomically" to the distance buffer will avoid that. All other sensors use this technique as well.